### PR TITLE
Provide tool to build and distribute deno-compiled binary on PyPi

### LINKED ISSUE #148: Upload cross-platform builds of deno bids-validator as .whl's to pypi
Not my idea... I think @effigies and/or @nellh had it: similarly to [uv](https://github.com/astral-sh/uv) which provides .whl files across platforms (see https://pypi.org/project/uv/0.5.26/#files), it would be nice to have bids-validator builds (pretty much output binary from `deno compile -ERNW -o bids-validator jsr:@bids/validator`) to be wrapped into a "full fledged" whl.

Anyone has some other project which does something like that?  I looked at `uv` but it seemed to rely on rust specific https://github.com/PyO3/maturin-action to build whls.

May be @jwodder have heard about anything or has good pointers for easy creation of whls for binaries?

### PR DESCRIPTION
**Pull request by candleindark**
This PR closes #148. Please see commit messages for details.

**Note**:
1. The `--allow-run` option is used to compiled the BIDS validator from local source. Without this option, the compilation result would require the user to grant permission to run subprocesses. This option was not used/needed when compiling from the published jsr package from some reason. 
2. For publishing to work, please set up a github secret `PYPI_TOKEN` containing the PyPi token.

TODOs:
- [x] created https://pypi.org/project/bids-validator-deno/ by uploading 2.0.3 "manual" build done on my @yarikoptic linux laptop . Required "faking" this PR branch to be on 2.0.3 tag for proper tagging. Not sure if we need/want to upload more wheels similarly -- this workflow I think would not work unless we make it force similar nastiness somehow
- [x] invited @effigies @tsalo @candleindark to maintain this PyPI project
- [x] BTW -- I also applied for BIDS community organization on PyPI FWIW  
- [x] add PYPI_TOKEN
- [x] remove mentioning of testing pypi
- [x] validate publishing on test PyPi with the final setup. https://test.pypi.org/project/bids-validator-deno/2.0.12/ 
- ~~set up publishing with OpenID Connect(POSTPONED)~~

### CODE REVIEW

File: .github/workflows/pypi-whl.yml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          path: tools/pypi-whl/dist/*.whl
+      - name: Publish wheels if corresponds to a tag
+        # TODO: might be suboptimal since we might skip a tag if not the latest commit
+        run: |
+            cd tools/pypi-whl
+            # if a clean tag, no - in it
+            if git describe | grep -e -; then 
+                echo not releasing; 
+            else
+                twine upload --verbose --repository testpypi dist/*.whl
```
Comment (yarikoptic) [inline .github/workflows/pypi-whl.yml:92]: potentially auth could be replaced by using "trusted repositories" setup on pypi.

File: .github/workflows/pypi-whl.yml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            ACTIVATE_PATH=.venv/bin/activate
+            BIDS_VALIDATOR_BINARY_NAME=bids-validator
+            VENV_BIN_DIR=.venv/bin
+          fi
+          source "$ACTIVATE_PATH"
+          pip install tools/pypi-whl/dist/*.whl
+          bids-validator --version
+          ls -lh "$VENV_BIN_DIR"
+          diff "$VENV_BIN_DIR"/"$BIDS_VALIDATOR_BINARY_NAME" tools/pypi-whl/dist/"$BIDS_VALIDATOR_BINARY_NAME"
+          deactivate
```
Comment (yarikoptic) [inline .github/workflows/pypi-whl.yml:77]: move this into a helper script e.g. `tools/pypi-whl/test-installation` -- too much for the workflow - makes it hard to read.

File: .github/workflows/pypi-whl.yml [hunk 3]
```diff
@@ -0,0 +1,96 @@
+name: PyPI Whl Publish
+
+on:
+  push:
+    branches:
+      - main
+      - deno-binary
```
Comment (yarikoptic) [inline .github/workflows/pypi-whl.yml:7]: eventually remove 
```suggestion
```

Comment (yarikoptic) [inline .github/workflows/pypi-whl.yml:7]: overall here aim for the same setup of `on` as in https://github.com/bids-standard/bids-validator/blob/main/.github/workflows/deno_tests.yml

```suggestion
on:
  push:
    branches: [main, dev]
    tags: ['*']
  pull_request:
    branches: [main, dev]
```

and then add corresponding `if` below for `twine` call:

File: .github/workflows/pypi-whl.yml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+          diff "$VENV_BIN_DIR"/"$BIDS_VALIDATOR_BINARY_NAME" tools/pypi-whl/dist/"$BIDS_VALIDATOR_BINARY_NAME"
+          deactivate
+      - name: Check if distribution's description will render correctly on PyPI
+        run: twine check tools/pypi-whl/dist/*.whl
+      - uses: actions/upload-artifact@v4
+        with:
+          name: wheel-${{ matrix.os }}
+          path: tools/pypi-whl/dist/*.whl
+      - name: Publish wheels if corresponds to a tag
+        # TODO: might be suboptimal since we might skip a tag if not the latest commit
```
Comment (yarikoptic) [inline .github/workflows/pypi-whl.yml:85]:
```suggestion
        if: github.ref_type == 'tag'
```

File: .github/workflows/pypi-whl.yml (hunk 1 above)
Comment (yarikoptic) [inline .github/workflows/pypi-whl.yml:92]: with above `if` we likely do not need this 

```suggestion
            twine upload --verbose --repository testpypi dist/*.whl
```

**Comment by effigies** [comment]
image

The `cp3*` is going to cause problems, as the package will only be installable with Python 3.9. Installing from test.pypi.org works because it finds `bids_validator_deno-2.0.7-py3-none-manylinux_2_17_x86_64.whl`.

You need to get back to the `py3-none-*`. I'm not sure how to do that in `pdm`. I could work it out in setuptools if I knew how to package a pre-built binary.

**Comment by yarikoptic** [comment]
right, somehow that got lost I guess -- may be pdm is not executed within the folder where that pdb_build.py is?

Please add testing that whl files have `-py3-none-` in their names!

**Comment by effigies** [comment]
Oh, I see. It reports using `cp39-cp39`, but then post-processing renames to `py3-none`.

**Comment by candleindark** [comment]
I tried to use the pdm frontend to invoke the build locally. It didn't work for me. Just use `python -m build` if want to try it locally, but the current setup is really for CI to do the building.

File: .github/workflows/pypi-whl.yml (hunk 2 above)
Comment (candleindark) [reply .github/workflows/pypi-whl.yml]: done

File: .github/workflows/pypi-whl.yml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        with:
+          name: wheel-${{ matrix.os }}
+          path: tools/pypi-whl/dist/*.whl
+      - name: Publish wheels if corresponds to a version tag
+        if: github.ref_type == 'tag'
+        working-directory: ../../
+        run: |
+          # Check if the tag is a version tag
+          if jq -e ".version[:1] != \"v\" and .version==\"$TAG\"" deno.json; then 
+              twine upload --verbose --repository testpypi dist/*.whl
```
Comment (yarikoptic) [inline .github/workflows/pypi-whl.yml:73]: for the final version should remove 
```suggestion
              twine upload --verbose dist/*.whl
```

**Comment by effigies** [comment]
Played with this a bit and wasn't entirely satisfied. Ended up spending way too much time on it. See #186.
